### PR TITLE
[CECO-570] Add operator introspection

### PR DESCRIPTION
**Pull request by khewonc**
### What does this PR do?

[BETA] Adds the ability to look at node details and create multiple daemonsets/EDSs with configs specific to a node's provider. Only GKE COS has been implemented so far.

```bash
$ kubectl get pods
NAME                                        READY   STATUS    RESTARTS   AGE
datadog-agent-gcp-cos-lt5tg                 4/4     Running   0          3d1h
datadog-agent-gcp-cos-r2r2r                 4/4     Running   0          3d1h
datadog-cluster-agent-67fd7dcfd9-qpln7      1/1     Running   0          3d1h
operator-datadog-operator-56fb4888c-jngxm   1/1     Running   0          3d1h
$ kubectl get ds
NAME                    DESIRED   CURRENT   READY   UP-TO-DATE   AVAILABLE   NODE SELECTOR   AGE
datadog-agent-gcp-cos   2         2         2       2            2           <none>          3d1h
```

### Motivation

What inspired you to submit this pull request?

### Additional Notes

This feature is in beta

### Describe your test plan

* Introspection disabled: the status section in the `kubectl describe` output of the `DatadogAgent` should include an `Agent List` section and the plugin should use the `Agent` information
* Introspection enabled (delete existing DatadogAgent to prevent stale DSs)
  * In a mixed cluster of GKE COS + non-COS nodes, there should be multiple DSs created: one for COS nodes and one for non-COS nodes (default)
  * Deleting all the nodes of a provider should cause the corresponding DS to be deleted and its status to be removed from the `Agent List`
  * plugin should show combined agent status

### CODE REVIEW

File: apis/datadoghq/v2alpha1/datadogagent_types.go [hunk 1]
```diff
@@ -1081,9 +1081,13 @@ type DatadogAgentStatus struct {
...
 	Conditions []metav1.Condition `json:"conditions"`
-	// The actual state of the Agent as an extended daemonset.
+	// The actual state of the Agent as a daemonset or an extended daemonset.
 	// +optional
-	Agent *commonv1.DaemonSetStatus `json:"agent,omitempty"`
+	// +listType=atomic
+	Agent []*commonv1.DaemonSetStatus `json:"agent,omitempty"`
+	// The combined actual state of the all Agent as daemonsets or extended daemonsets.
+	// +optional
+	CombinedAgent *commonv1.DaemonSetStatus `json:"combinedAgent,omitempty"`
```
Comment (celenechang) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:1090]: how about `AgentSummary` ?

File: controllers/datadogagent_controller.go [hunk 2]
```diff
@@ -42,6 +42,7 @@ type DatadogAgentReconciler struct {
 	client.Client
 	VersionInfo  *version.Info
 	PlatformInfo kubernetes.PlatformInfo
+	Providers    *kubernetes.ProviderStore
```
Comment (celenechang) [inline controllers/datadogagent_controller.go:45]: could we call it `ProviderStore` instead?

**Review by celenechang — COMMENTED** [review]
feel free to address these after handling the subtasks!

**Review by celenechang — APPROVED** [review]

File: controllers/datadogagent/controller_reconcile_agent.go [hunk 3]
```diff
@@ -90,24 +123,39 @@ func (r *Reconciler) reconcileV2Agent(logger logr.Logger, requiredComponents fea
...
+		if r.options.IntrospectionEnabled {
+			agentNameWithProvider := kubernetes.GetAgentNameWithProvider(daemonset.Name, provider, co.Name)
+			co.Name = &agentNameWithProvider
+		}
+		componentOverride = co
+	} else {
+		if r.options.IntrospectionEnabled {
+			overrideFromProvider := kubernetes.ComponentOverrideFromProvider(daemonset.Name, provider)
+			componentOverride = &overrideFromProvider
+		}
```
Comment (CharlyF) [inline controllers/datadogagent/controller_reconcile_agent.go:148]:
```suggestion
		if r.options.IntrospectionEnabled {
			overrideFromProvider := kubernetes.ComponentOverrideFromProvider(daemonset.Name, provider)
			componentOverride = &overrideFromProvider
		} else {
		        daemonset.Labels[apicommon.MD5AgentDeploymentProviderLabelKey] =  kubernetes.LegacyProvider
		}
```

**Review by CharlyF — APPROVED** [review]
2 nits, 1 improvement

File: controllers/datadogagent/controller_reconcile_agent.go [hunk 4]
```diff
@@ -50,19 +57,34 @@ func (r *Reconciler) reconcileV2Agent(logger logr.Logger, requiredComponents fea
...
+				agentNameWithProvider := kubernetes.GetAgentNameWithProvider(eds.Name, provider, co.Name)
+				co.Name = &agentNameWithProvider
+			}
+			componentOverride = co
+		} else {
+			if r.options.IntrospectionEnabled {
+				overrideFromProvider := kubernetes.ComponentOverrideFromProvider(eds.Name, provider)
+				componentOverride = &overrideFromProvider
+			}
+		}
```
Comment (CharlyF) [inline controllers/datadogagent/controller_reconcile_agent.go:78]:
```suggestion
		} else {
			if r.options.IntrospectionEnabled {
				overrideFromProvider := kubernetes.ComponentOverrideFromProvider(eds.Name, provider)
				componentOverride = &overrideFromProvider
			} else {
		        eds.Labels[apicommon.MD5AgentDeploymentProviderLabelKey] =  kubernetes.LegacyProvider
		}
			
		}
```

**Pull request merged**